Rubberduck.UnitTesting/UnitTesting/TestEngine.cs

```diff
@@ -158,6 +160,103 @@ public void Run(IEnumerable<TestMethod> tests)
                 RunInternal(queued);
             });
         }
+        public string RunWithResults(IEnumerable<TestMethod> tests)
+        {
+            if (tests == null)
+            {
+                // Trigger the ParseRequest programmatically to make the Tests property available.
+                var parseCompletion = new TaskCompletionSource<bool>();
+                EventHandler<ParserStateEventArgs> parseCompletedHandler = null;
+
+                parseCompletedHandler = (sender, args) =>
+                {
+                    if (args.State == ParserState.Ready)
+                    {
+                        _state.StateChanged -= parseCompletedHandler; // Unsubscribe from the event
+                        parseCompletion.SetResult(true); // Signal that parsing is complete
+                    }
+                };
+
+                _state.StateChanged += parseCompletedHandler;
+                _state.OnParseRequested(this);
+
+                parseCompletion.Task.Wait();
+
+                tests = Tests;
+            }
+
+            var queued = tests.ToList();
+            var results = new List<TestInfo>();
+
+            foreach (var test in queued.Where(item => _knownOutcomes.ContainsKey(item)))
+            {
+                _knownOutcomes.Remove(test);
+            }
+
+            Task.Run(() =>
+            {
+                var suspensionResult = _state.OnSuspendParser(this, AllowedRunStates, () =>
+                {
+                    results.AddRange(RunWhileSuspendedWithResults<TestInfo>(tests));
+                });
+
+                switch (suspensionResult.Outcome)
+                {
+                    case SuspensionOutcome.Completed:
+                        break;
+                    case SuspensionOutcome.Canceled:
+                        Logger.Debug("Test execution canceled.");
+                        break;
+                    default:
+                        Logger.Warn($"Test execution failed with suspension outcome {suspensionResult.Outcome}.");
+                        if (suspensionResult.EncounteredException != null)
+                        {
+                            Logger.Error(suspensionResult.EncounteredException);
+                        }
+                        break;
+                }
+            }).GetAwaiter().GetResult(); // Ensure the task completes before returning results.
+
+            var resultBuilder = new StringLineBuilder();
+            foreach (var result in results)
+            {
+                // Get the TestName, but stop at the first \r\n to get only the signature
+                int index = result.TestName.IndexOf("\r\n");
+                var signature = index >= 0 ? result.TestName.Substring(0, index) : result.TestName;
+                resultBuilder.AppendLine($"{result.Result.Outcome}: {signature}");
```

Comment:
I'm not sure why the signature is relevant, seems to me just the name is sufficient (besides they're all going to be `Public Sub` procedures). 

Thinking the results should probably be output as a json string, and let the client display it the way they want it, no?

Reply:
Yeah, the name would be sufficient. Is there a way to get the name of the procedure from a `TestResult`? I had to create the struct `TestInfo` as a way to get something close to the method name, but then we need to parse it to get only the name.